torch/csrc/jit/script/tree.h

```diff
@@ -102,10 +99,36 @@ struct Tree : c10::intrusive_ptr_target {
       throw std::runtime_error(ss.str());
     }
   }
-  virtual ~Tree() = default;
+
+  void *operator new(size_t, void* p_) { return p_; }
+  void *operator new(size_t size) {
+    if (freelist_.empty() || size + slaboffset_ > ALLOC_SIZE) {
```

Comment:
This isn't thread safe. If we have a parallel predictor loading models at the same time there is no way to organize the arena load/stores to avoid freeing the other threads memory.

Reply:
Do you think having an arena per thread would suffice?